I tried to check for C++ compiler warnings in internal Python C headers, but I got many C++ compiler errors in `mimalloc` and I don't know how to handle them. Examples:

```c
In file included from /usr/include/features.h:524,
                 from /usr/include/assert.h:35,
                 from /home/vstinner/python/main/Include/Python.h:19,
                 from extension.cpp:13:
/home/vstinner/python/main/Include/internal/mimalloc/mimalloc/types.h:73:33: error: expected unqualified-id before 'noexcept'
   73 | #define mi_decl_throw           __THROW
      |                                 ^~~~~~~
/home/vstinner/python/main/Include/internal/mimalloc/mimalloc/types.h:611:31: note: in expansion of macro 'mi_decl_throw'
  611 | mi_decl_noreturn mi_decl_cold mi_decl_throw
      |                               ^~~~~~~~~~~~~
```

and

```c
In file included from /home/vstinner/python/main/Include/internal/pycore_mimalloc.h:44,
                 from /home/vstinner/python/main/Include/internal/pycore_tstate.h:13,
                 from /home/vstinner/python/main/Include/internal/pycore_pystate.h:13,
                 from extension.cpp:20:
/home/vstinner/python/main/Include/internal/mimalloc/mimalloc/internal.h: In function 'bool _mi_is_aligned(void*, size_t)':
/home/vstinner/python/main/Include/internal/mimalloc/mimalloc/types.h:613:49: error: '_mi_assert_fail' was not declared in this scope; did you mean '__assert_fail'?
  613 | #define mi_assert(expr)     ((expr) ? (void)0 : _mi_assert_fail(#expr,__FILE__,__LINE__,__func__))
      |                                                 ^~~~~~~~~~~~~~~
/home/vstinner/python/main/Include/internal/mimalloc/mimalloc/types.h:619:31: note: in expansion of macro 'mi_assert'
  619 | #define mi_assert_internal    mi_assert
      |                               ^~~~~~~~~
/home/vstinner/python/main/Include/internal/mimalloc/mimalloc/internal.h:273:3: note: in expansion of macro 'mi_assert_internal'
  273 |   mi_assert_internal(alignment != 0);
      |   ^~~~~~~~~~~~~~~~~~
```

--- a/Include/internal/pycore_interp_structs.h
+++ b/Include/internal/pycore_interp_structs.h
@@ -129,8 +129,6 @@ struct _atexit_runtime_state {
 //###################
 // interpreter atexit
 
-typedef void (*atexit_datacallbackfunc)(void *);
-
 typedef struct atexit_callback {
     atexit_datacallbackfunc func;
     void *data;

--- a/Lib/test/test_cext/setup.py
+++ b/Lib/test/test_cext/setup.py
@@ -36,16 +36,23 @@ if not support.MS_WINDOWS:
             # with the -Werror=declaration-after-statement compiler flag.
             '-Werror=declaration-after-statement',
         )
+    INTERNAL_CFLAGS = [*BASE_CFLAGS]
 else:
     # MSVC compiler flags
     BASE_CFLAGS = [
-        # Display warnings level 1 to 4
-        '/W4',
         # Treat all compiler warnings as compiler errors
         '/WX',
     ]
-    PUBLIC_CFLAGS = [*BASE_CFLAGS]
-INTERNAL_CFLAGS = [*BASE_CFLAGS]
+    PUBLIC_CFLAGS = [
+        *BASE_CFLAGS,
+        # Display warnings level 1 to 4
+        '/W4',
+    ]
+    INTERNAL_CFLAGS = [
+        *BASE_CFLAGS,
+        # Display warnings level 1 to 3
+        '/W3',
+    ]
 
 
 def main():

--- a/Lib/test/test_cppext/__init__.py
+++ b/Lib/test/test_cppext/__init__.py
@@ -24,7 +24,7 @@ SETUP = os.path.join(os.path.dirname(__file__), 'setup.py')
 @support.requires_venv_with_pip()
 @support.requires_subprocess()
 @support.requires_resource('cpu')
-class TestCPPExt(unittest.TestCase):
+class BaseTests:
     def test_build(self):
         self.check_build('_testcppext')
 
@@ -33,10 +33,6 @@ class TestCPPExt(unittest.TestCase):
         # in practice we do maintain C++03 compatibility in public headers.
         # Please ask the C API WG before adding a new C++11-only feature.
         self.check_build('_testcpp03ext', std='c++03')
-
-    @support.requires_gil_enabled('incompatible with Free Threading')
-    def test_build_limited_cpp03(self):
-        self.check_build('_test_limited_cpp03ext', std='c++03', limited=True)
 
     @unittest.skipIf(support.MS_WINDOWS, "MSVC doesn't support /std:c++11")
     def test_build_cpp11(self):
@@ -47,10 +43,6 @@ class TestCPPExt(unittest.TestCase):
     @unittest.skipIf(not support.MS_WINDOWS, "need Windows")
     def test_build_cpp14(self):
         self.check_build('_testcpp14ext', std='c++14')
-
-    @support.requires_gil_enabled('incompatible with Free Threading')
-    def test_build_limited(self):
-        self.check_build('_testcppext_limited', limited=True)
 
     def check_build(self, extension_name, std=None, limited=False):
         venv_dir = 'env'
@@ -109,6 +101,20 @@ class TestCPPExt(unittest.TestCase):
                '-X', 'showrefcount',
                '-c', f"import {extension_name}"]
         run_cmd('Import', cmd)
+
+
+class TestPublicCAPI(BaseTests, unittest.TestCase):
+    @support.requires_gil_enabled('incompatible with Free Threading')
+    def test_build_limited_cpp03(self):
+        self.check_build('_test_limited_cpp03ext', std='c++03', limited=True)
+
+    @support.requires_gil_enabled('incompatible with Free Threading')
+    def test_build_limited(self):
+        self.check_build('_testcppext_limited', limited=True)
+
+
+class TestInteralCAPI(BaseTests, unittest.TestCase):
+    TEST_INTERNAL_C_API = True
 
 
 if __name__ == "__main__":

--- a/Lib/test/test_cppext/extension.cpp
+++ b/Lib/test/test_cppext/extension.cpp
@@ -6,7 +6,16 @@
 // Always enable assertions
 #undef NDEBUG
 
+#ifdef TEST_INTERNAL_C_API
+#  define Py_BUILD_CORE 1
+#endif
+
 #include "Python.h"
+
+#ifdef TEST_INTERNAL_C_API
+   // gh-135906: Check for compiler warnings in the internal C API
+#  include "internal/pycore_frame.h"
+#endif
 
 #ifndef MODULE_NAME
 #  error "MODULE_NAME macro must be defined"

--- a/Lib/test/test_cppext/setup.py
+++ b/Lib/test/test_cppext/setup.py
@@ -47,6 +47,7 @@ def main():
     std = os.environ.get("CPYTHON_TEST_CPP_STD", "")
     module_name = os.environ["CPYTHON_TEST_EXT_NAME"]
     limited = bool(os.environ.get("CPYTHON_TEST_LIMITED", ""))
+    internal = bool(int(os.environ.get("TEST_INTERNAL_C_API", "0")))
 
     cppflags = list(CPPFLAGS)
     cppflags.append(f'-DMODULE_NAME={module_name}')
@@ -81,6 +82,9 @@ def main():
     if limited:
         version = sys.hexversion
         cppflags.append(f'-DPy_LIMITED_API={version:#x}')
+
+    if internal:
+        cppflags.append('-DTEST_INTERNAL_C_API=1')
 
     # On Windows, add PCbuild\amd64\ to include and library directories
     include_dirs = []

